### ISSUE TITLE
ChatDataset mid training resume not reproducible with multi epoch

### ISSUE THREAD

**Issue by slimfrkha** [pytorch/torchtitan#2864]
### Bug description
Shuffling the dataset is different between
1. training relooping
2. state dict loading to resume training
This problem appears on ChatDataset in multi epoch training if resuming after epoch > 1

Basically the below is happening right now:
```python
from datasets import load_dataset
ds = load_dataset("cornell-movie-review-data/rotten_tomatoes", split="validation")

# in training
shuffled_ds_training = ds
for epoch in range(4):
        shuffled_ds_training =shuffled_ds_training.shuffle(seed=42 + epoch)
print(shuffled_ds_training['label'][:10])

# when loading from state to resume training
shuffled_ds_resume_training = ds.shuffle(seed=42 + epoch)
print(shuffled_ds_resume_training['label'][:10])
```

## Solution
Either on state dict load, do a perform to replay all the previous epochs shuffling or dissociate shuffled data from original dataset and shuffle always from original dataset.

### Versions

main branch latest commit `0e0590c137599276d36128abc1702efe9e091607` as of this writing.
